Revealing preimages 2

hewison-chris opened the pull request:
Same as PR https://github.com/bosagora/agora/pull/3153 without trivial logging updates.

hewison-chris committed:
Allow preimage reveal beyond enrollemnt

If the pre-images are only revealed up to the end of the enrollment
cycle then there is more risk for the validator to get slashed in the
first block of the new cycle.

hewison-chris committed:
Do not update stored preimage if we have higher

hewison-chris committed:
Do not reset validator enroll_key

We can still use the stored `enroll_key` to send valid preimages if the
timer fires before we re-enroll. This will improve preimage revealing as
we will still share it with other nodes whilst in the process of
re-enrolling.

hewison-chris committed:
Ignore envelopes that originate from self

We can return early once we know that this envelope originates from this
node. Also we avoid `SCP indicated invalid envelope` in logs which is
only because SCP detects it is from current node.

hewison-chris committed:
Ignore envelope for height after next block

Also trigger the catchup task as we are most likely behind the other
nodes in block height.

hewison-chris committed:
Only re-arm task timers if not pending

hewison-chris committed:
Test code: using enrollments from test case

In some tests not all the nodes will have pre-images at a certain
height. In these tests, rather than relying on the block header at the
end of the enrollment cycle, we can pass the enrollments of those nodes
that should have a preimage at that height.

hewison-chris committed:
Use round timeout once nominating

When nomination is started then use the computed round timeout to also
trigger the next nomination attempt.

linked0 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -473,32 +468,16 @@ public class EnrollmentManager
-        @safe
+    public PreImageInfo getPreimage (in Height height) @safe
     {
-        const enrolled = this.validator_set.getEnrolledHeight(height, this.enroll_key);
-        if (enrolled == ulong.max)
```
Can we leave these lines with some log like "The enrollment doesn't exist or is out of the cycle" for a while? It's because we can find out how many times the situation happens and how it does. Anyway, the situation where `getPreimage` is called without enrollment is not so normal situation, I think.

hewison-chris replied:
As we are revealing into the next cycle then we will not yet be enrolled so it is expected behavior. The idea is that we reveal before the enrollment as if we do not enroll then it is not a problem that we revealed. However if we enroll and the `preimage` is not revealed in time we get slashed which is a big problem.

linked0 replied:
Ok, I got it. But my remaining question is why there was a situation in the `ValidatorRecurringEnrollment` test, where no pre-images of a validator are revealed on other validators although a validator reveals preimages right after accepting new blocks. It comes to the mind that the `preimage_reveal_interval` should have been shorter than `nomination_interval` or sth.

hewison-chris replied:
I think the problem there was more related to the fact the peers changed after the `preimages` had been revealed. However the `preimage` catchup solves that.

linked0 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -156,6 +156,9 @@ public extern (C++) class Nominator : SCPDriver
     private const Duration nomination_interval;
 
+    /// User configured nomination frequency
+    private Duration round_timeout;
```
If we use this new variable for nomination timing, we should remove the `nomination_interval` variable. There's no use for the old variable as far as I can see.

hewison-chris replied:
We need it to reset the `round_timeout` to the configured value from `config.validator.nomination_interval` which is passed into the `Nominator` class when constructed.

linked0 reviewed (commented):
Broadly LGTM. Two comments.

hewison-chris committed:
Send envleopes again if not new nomination

This will cause a large volume of envelope gossiping so will need to be
optimized to only happen when we are not making progress. But let's see
if this actually helps in ci first.

linked0 reviewed (approved):
LGTM

hewison-chris committed:
Integration test: Reduce nomination interval

To prevent lots of empty blocks with a block interval of 5 seconds we
need to check if it is time to nominate more frequently.

hewison-chris committed:
Correct the log for rejected pre-image

It was writing the pre-image hash where it should be the utxo

hewison-chris committed:
Check that utxo is valid in pre-image

Only send pre-image if the utxo is set.
